File: sesion1/benchmark/mul_benchmark.c
#include "mul_functions.h"

#define ROW_MAJOR_STR "Row-major order"
#define COLUMN_MAJOR_STR "Column-major order"
#define Z_ORDER_STR "Z order"

/**
 * Struct to hold the results of the benchmark
 */
typedef struct
{
    double row_major_time;    /* Execution time with row major algorithm */
    double column_major_time; /* Execution time with column major algorithm */
    double *zorder_times;     /* Execution time with z-order algorithm */
} Result;

/**
 * Prints the results of the benchmark
 * @param matrix_size Size of the matrices
 * @param num_blocks Number of blocks for z-order algorithm
 * @param block_sizes Array of block sizes
 * @param result Pointer to the result struct
 */
void print_results(int matrix_size, int num_blocks, int *block_sizes, Result *result)
{
    printf("%d;%s;%s;%f\n", matrix_size, "-", "row", result->row_major_time);
    printf("%d;%s;%s;%f\n", matrix_size, "-", "col", result->column_major_time);
    for (int i = 0; i < num_blocks; i++)
    {
        printf("%d;%d;%s;%f\n", matrix_size, block_sizes[i], "zor", result->zorder_times[i]);
    }
}

/**
 * Calculates the average execution times for the benchmark
 * @param result Pointer to the result struct
 * @param iterations Number of iterations
 * @param num_blocks Number of blocks for z-order algorithm
 */
void calc_results(Result *result, int iterations, int num_blocks)
{
    result->row_major_time /= iterations;
    result->column_major_time /= iterations;
    for (int i = 0; i < num_blocks; i++)
    {
        result->zorder_times[i] /= iterations;
    }
}

/**
 * Runs the benchmark for the given matrices and block sizes
 * It fills the matrices with zeros, runs the row major, column major, and z-order algorithms,
 * checks the results, and stores the execution times
 * @param matrix_size Size of the matrices
 * @param block_sizes Array of block sizes
 * @param num_blocks Number of blocks for z-order algorithm
 * @param result Pointer to the result struct
 * @param A First matrix
 * @param B Second matrix
 * @param C_rows Result matrix for row major algorithm
 * @param C_columns Result matrix for column major algorithm
 * @param C_zorder Result matrix for z-order algorithm
 */
void benchmark(int matrix_size, int *block_sizes, int num_blocks, Result *result,
               int **A, int **B, int **C_rows, int **C_columns, int **C_zorder)
{
    clock_t start, end;

    // Initialize matrices with zeros
    fill_matrix(matrix_size, matrix_size, C_rows, 0);
    fill_matrix(matrix_size, matrix_size, C_columns, 0);

    // Row major benchmark
    start = clock();
    row_major_mul(matrix_size, matrix_size, A, B, C_rows);
    end = clock();
    result->row_major_time += (double)(end - start) / CLOCKS_PER_SEC;

    // Column major benchmark
    start = clock();
    column_major_mul(matrix_size, matrix_size, A, B, C_columns);
    end = clock();
    result->column_major_time += (double)(end - start) / CLOCKS_PER_SEC;

    // Check row major and column major results
    if (!check_matrices(matrix_size, C_rows, C_columns))
    {
        fprintf(stderr, "Error: Results do not match.\n");
        exit(1);
    }

    // Z-order benchmark
    for (int i = 0; i < num_blocks; i++)
    {
        // Initialize z-order matrix with zeros
        fill_matrix(matrix_size, matrix_size, C_zorder, 0);

        // Z-order multiplication
        start = clock();
        zorder_mul(matrix_size, matrix_size, A, B, C_zorder, block_sizes[i]);
        end = clock();
        result->zorder_times[i] += (double)(end - start) / CLOCKS_PER_SEC;

        // Check z-order results
        if (!check_matrices(matrix_size, C_rows, C_zorder))
        {
            fprintf(stderr, "Error: Results do not match for z-order multiplication.\n");
            exit(1);
        }
    }
}

/**
 * Initializes the result struct
 * @param num_blocks Number of blocks for z-order algorithm
 * @return Initialized result struct
 */
Result init_results(int num_blocks)
{
    Result result = {0.0, 0.0, (double *)malloc(num_blocks * sizeof(double))};
    if (result.zorder_times == NULL)
    {
        fprintf(stderr, "Error: Could not allocate memory for z-order times.\n");
        exit(1);
    }

    for (int i = 0; i < num_blocks; i++)
    {
        result.zorder_times[i] = 0.0;
    }

    return result;
}

/**
 * Calculates the block sizes for the given matrix size
 * It is used to determine the block sizes for the z-order algorithm
 * @param matrix_size Size of the matrix
 * @param num_block_sizes Pointer to the number of block sizes
 * @return Array of block sizes
 */
int *calculate_block_sizes(int matrix_size, int *num_block_sizes)
{
    int *block_sizes = (int *)malloc(matrix_size * sizeof(int));
    if (block_sizes == NULL)
    {
        fprintf(stderr, "Error: Could not allocate memory for block sizes.\n");
        exit(1);
    }

    int count = 0;
    block_sizes[count++] = 1;
    for (int i = 2; i < matrix_size; i++)
    {
        if (matrix_size % i == 0)
        {
            block_sizes[count++] = i;
        }
    }
    block_sizes[count++] = matrix_size;

    *num_block_sizes = count;

    return block_sizes;
}

/**
 * Runs the benchmark for the given matrix sizes and iterations
 * It initializes the matrices, runs the benchmark, and prints the results
 * @param matrix_sizes Array of matrix sizes
 * @param size_count Number of matrix sizes
 * @param iterations Number of iterations to run the benchmark
 */
void run_benchmark(int matrix_sizes[], int size_count, int iterations)
{
    for (int i = 0; i < size_count; i++)
    {
        int **A = NULL, **B = NULL, **C_rows = NULL, **C_columns = NULL, **C_zorder = NULL;
        int matrix_size = matrix_sizes[i];
        int num_blocks;

        int *block_sizes = calculate_block_sizes(matrix_size, &num_blocks);
        Result result = init_results(num_blocks);
        init_matrices(matrix_size, &A, &B, &C_rows, &C_columns, &C_zorder);

        for (int j = 0; j < iterations; j++)
        {
            benchmark(matrix_size, block_sizes, num_blocks, &result, A, B, C_rows, C_columns, C_zorder);
        }

        calc_results(&result, iterations, num_blocks);
        print_results(matrix_size, num_blocks, block_sizes, &result);

        free(block_sizes);
        free(result.zorder_times);
        free_matrices(matrix_size, A, B, C_rows, C_columns, C_zorder);
    }
}

/**
 * Main function
 */
int main()
{
    clock_t start, end;
    int matrix_sizes[] = {2, 4, 8, 16, 32, 64, 128, 256, 398, 512, 636, 774, 892, 1024, 1152, 1280, 1408, 1536};
    int iterations = 32;

    start = clock();
    printf("Matrix size;Block size;Algorithm;Time(s)\n");
    run_benchmark(matrix_sizes, sizeof(matrix_sizes) / sizeof(matrix_sizes[0]), iterations);
    end = clock();

    printf("\nTotal execution time: %f seconds\n", (double)(end - start) / CLOCKS_PER_SEC);
    printf("Iterations: %d\n", iterations);
    printf("Benchmark completed.\n");

    return 0;
}

File: sesion2/benchmark/mul_benchmark.py
import time
import sys

sys.path.append("..")

from multiply_matrices import run_phase_1_row_col, run_phase_1_zorder
from multiply_matrices_hybrid import run_phase_2_row_col, run_phase_2_zorder
from multiply_matrices_hybrid_pro import run_phase_3_row_col, run_phase_3_zorder
from typing import List, Dict

row_major_str = "Row-major order"
column_major_str = "Column-major order"
z_order_str = "Z order"

def run_phase(phase_id: int, matrix_sizes: List[int], iterations: int):
    """Runs a phase of the experiment."""
    results = {}
    
    print("Matrix size;Block size;Phase;Row-major order (s);Column-major order (s);Z order (s)")
    
    for matrix_size in matrix_sizes:
        block_sizes = calculate_block_sizes(matrix_size)
        initialize_results(results, matrix_size, block_sizes)
        process_block_sizes(phase_id, matrix_size, block_sizes, iterations, results)
        print_results(phase_id, matrix_size, iterations, results)

def calculate_block_sizes(matrix_size: int) -> List[int]:
    """Calculate the block sizes for the given matrix size."""
    block_sizes = []
    for block_size in range(1, matrix_size + 1):
        if matrix_size % block_size == 0 and matrix_size // block_size >= 2:
            block_sizes.append(block_size)
    block_sizes.append(matrix_size)
    return block_sizes

def initialize_results(results: Dict[str, float], matrix_size: int, block_sizes: List[int]):
    """Initialize the results dictionary."""
    if matrix_size not in results:
        results[matrix_size] = {
            row_major_str: 0.0,
            column_major_str: 0.0,
            z_order_str: {block_size: 0 for block_size in block_sizes}
        }

def process_block_sizes(phase_id: int, matrix_size: int, block_sizes: List[int], iterations: int, results: Dict[str, float]):
    """Process the block sizes for the given matrix size."""
    for _ in range(iterations):
        if phase_id == 1:
            row_time, col_time = run_phase_1_row_col(matrix_size)
        elif phase_id == 2:
            row_time, col_time = run_phase_2_row_col(matrix_size)
        else:
            row_time, col_time = run_phase_3_row_col(matrix_size)
        
        results[matrix_size][row_major_str] += row_time
        results[matrix_size][column_major_str] += col_time
        for block_size in block_sizes:
            if phase_id == 1:
                zorder_time = run_phase_1_zorder(matrix_size, block_size)
            elif phase_id == 2:
                zorder_time = run_phase_2_zorder(matrix_size, block_size)
            else:
                zorder_time = run_phase_3_zorder(matrix_size, block_size)
            results[matrix_size][z_order_str][block_size] += zorder_time

def print_results(phase_id: int, matrix_size: int, iterations: int, results: Dict[str, float]):
    """Print the results stored in the dictionary."""
    row_major_avg = results[matrix_size][row_major_str] / iterations
    column_major_avg = results[matrix_size][column_major_str] / iterations

    for block_size, time in results[matrix_size][z_order_str].items():
        zorder_avg = time / iterations
        print(f"{matrix_size};{block_size};{phase_id};{row_major_avg:f};{column_major_avg:f};{zorder_avg:f}")

if __name__ == "__main__":
    matrix_sizes = [2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]
    iterations = 32
    
    start = time.time()
    run_phase(1, matrix_sizes, iterations)
    run_phase(2, matrix_sizes, iterations)
    run_phase(3, matrix_sizes, iterations)
    end = time.time()
    
    print(f"\nTotal execution time: {end - start:f} seconds")